### PR TITLE
New tracker API - stream managers registration summary

### PR DESCRIPTION
**Pull request by objmagic**
resolves https://github.com/twitter/heron/issues/1627

New tracker API ``/topologies/stmgrsregsummary?cluster=<cluster>&topology=<topology>&environ=<environment>`` that shows stream managers registration summary info. 

Schema of the result is 
```
{'registered_stmgrs': list of registered stmgrs names,
 'absent_stmgrs': list of absent stmgrs names}
```

### CODE REVIEW

File: heron/tmaster/src/cpp/manager/stats-interface.cpp [hunk 1]
```diff
@@ -137,6 +143,37 @@ void StatsInterface::HandleExceptionSummaryRequest(IncomingHTTPRequest* _request
...
+    _request->ExtractFromPostData(0, _request->GetPayloadSize());
+  heron::proto::tmaster::StmgrsRegistrationSummaryRequest stmgrs_reg_request;
+  if (!stmgrs_reg_request.ParseFromArray(request_data, _request->GetPayloadSize())) {
+    LOG(ERROR) << "Unable to deserialize post data specified in" <<
+      "StmgrsRegistrationSummaryRequest" << std::endl;
+    http_server_->SendErrorReply(_request, 400);
+    delete _request;
+    return;
+  }
+  LOG(INFO) << "OK GOT STMGRS REQUEST!!!";
```
Comment (kramasamy) [inline heron/tmaster/src/cpp/manager/stats-interface.cpp]: Remove extra log information that you introduced for debugging.

File: heron/tmaster/src/cpp/manager/stats-interface.cpp [hunk 2]
```diff
@@ -137,6 +143,37 @@ void StatsInterface::HandleExceptionSummaryRequest(IncomingHTTPRequest* _request
...
+  auto stmgrs_reg_summary_response = tmaster_->GetStmgrsRegSummary();
+  LOG(INFO) << "GOT RESPONSE!!!";
```
Comment (kramasamy) [inline heron/tmaster/src/cpp/manager/stats-interface.cpp]: ditto

**Comment by huijunw** [comment]
This feature sits along with "/stats","/exceptions","/exceptionsummary" but is not metircs/exceptions stats related. It will remain in TMaster when we eventually move tmetrics-collector to MetricsCache?

**Comment by objmagic** [comment]
@huijunw yes, it will remain in TMaster. As for the URL path, do you have any suggestion? Is `/topologies/stmgrsregsummary` good?

**Comment by huijunw** [comment]
The URL looks good to me. @billonahill may have ideas on path naming though.

**Comment by billonahill** [comment]
This is a great feature, but as implemented it seems overly specific. Could we instead extend upon `/topologies/executionstate` (I don't recall exactly what that returns) or introduce some other endpoint that represents topology state? This seems more extensible.

I would then expect to be able to make a single request to find the state of the topology, along with a collection of objects for all stream managers, each including their registration state.

That way the API would have a few key concepts like physical plan, logical plan and state.

**Comment by objmagic** [comment]
ping for review @billonahill @huijunw

File: heron/tools/tracker/src/python/handlers/metadatahandler.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+class MetaDataHandler(BaseHandler):
+  """
+  URL - /topologies/metadata
+  Parameters:
+   - cluster (required)
+   - environ (required)
+   - role - (optional) Role used to submit the topology.
+   - topology (required) name of the requested topology
+
+  The response JSON is a dictionary with all the
```
Comment (billonahill) [inline heron/tools/tracker/src/python/handlers/metadatahandler.py]: please elaborate on what should and should not be included in /metadata

File: heron/tools/tracker/src/python/handlers/metadatahandler.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    try:
+      cluster = self.get_argument_cluster()
+      role = self.get_argument_role()
+      environ = self.get_argument_environ()
+      topology_name = self.get_argument_topology()
+      topology_info = self.tracker.getTopologyInfo(topology_name, cluster, role, environ)
+      metadata = topology_info["metadata"]
+      self.write_success_response(metadata)
+    except Exception as e:
+      Log.debug(traceback.format_exc())
```
Comment (billonahill) [inline heron/tools/tracker/src/python/handlers/metadatahandler.py]: prepend something contextual log string before outputting a stack trace

File: heron/tools/tracker/src/python/handlers/runtimestatehandler.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+class RuntimeStateHandler(BaseHandler):
+  """
+  URL - /topologies/runtimestate
+  Parameters:
+   - cluster (required)
+   - environ (required)
+   - role - (optional) Role used to submit the topology.
+   - topology (required) name of the requested topology
+
+  The response JSON is a dictionary with all the
```
Comment (billonahill) [inline heron/tools/tracker/src/python/handlers/runtimestatehandler.py]: please elaborate on what should and should not be included in /runtimestate

File: heron/tools/tracker/src/python/handlers/runtimestatehandler.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    Get summary of stream managers registration summary
+    """
+    if not tmaster or not tmaster.host or not tmaster.stats_port:
+      return
+    reg_request = tmaster_pb2.StmgrsRegistrationSummaryRequest()
+    request_str = reg_request.SerializeToString()
+    port = str(tmaster.stats_port)
+    host = tmaster.host
+    url = "http://{0}:{1}/stmgrsregistrationsummary".format(host, port)
+    Log.debug("Creating request object.")
```
Comment (billonahill) [inline heron/tools/tracker/src/python/handlers/runtimestatehandler.py]: this debug message doesn't seem useful

File: heron/tools/tracker/src/python/handlers/stmgrsregsummaryhandler.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import tornado.web
+
+from heron.common.src.python.utils.log import Log
+from heron.proto import tmaster_pb2
+from heron.tools.tracker.src.python.handlers import BaseHandler
+
+# pylint: disable=attribute-defined-outside-init
+class StmgrsRegSummaryHandler(BaseHandler):
+  """
+  URL - /topologies/stmgrsregsummary?cluster=<cluster>&topology=<topology> \
```
Comment (billonahill) [inline heron/tools/tracker/src/python/handlers/stmgrsregsummaryhandler.py]: Why do we still need this endpoint? I thought the plan was to not add it, but instead use /runtimestate to serve this info. Can we remove this?

File: heron/tmaster/src/cpp/manager/stats-interface.cpp [hunk 8]
```diff
@@ -137,6 +143,33 @@ void StatsInterface::HandleExceptionSummaryRequest(IncomingHTTPRequest* _request
...
+    return;
+  }
+  auto stmgrs_reg_summary_response = tmaster_->GetStmgrsRegSummary();
+  sp_string response_string;
+  CHECK(stmgrs_reg_summary_response->SerializeToString(&response_string));
+  OutgoingHTTPResponse* http_response = new OutgoingHTTPResponse(_request);
+  http_response->AddHeader("Content-Type", "application/octet-stream");
+  std::ostringstream length_str;
+  length_str << response_string.size();
+  http_response->AddHeader("Content-Length", length_str.str());
```
Comment (objmagic) [inline heron/tmaster/src/cpp/manager/stats-interface.cpp]: `std::string(length_str)`? Fails to compile using Clang (Apple LLVM version 8.0.0 (clang-800.0.38)).

Error message:
```
ERROR: .../heron/tmaster/src/cpp/BUILD:3:1: C++ compilation of rule '//heron/tmaster/src/cpp:tmaster-cxx' failed: cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 60 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
heron/tmaster/src/cpp/manager/stats-interface.cpp:165:46: error: no matching conversion for functional-style cast from 'std::ostringstream' (aka 'basic_ostringstream<char>') to 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >')
  http_response->AddHeader("Content-Length", std::string(length_str));
```

Comment (congwang) [inline heron/tmaster/src/cpp/manager/stats-interface.cpp]: **to_**string()

Comment (objmagic) [inline heron/tmaster/src/cpp/manager/stats-interface.cpp]: oh my mad. definitely typo

**Comment by objmagic** [comment]
I have updated.

**Comment by congwang** [comment]
The C++ part looks good to me.

File: heron/tools/tracker/src/python/main.py [hunk 9]
```diff
@@ -55,6 +55,8 @@ def __init__(self):
         (r"/topologies/executionstate", handlers.ExecutionStateHandler, {"tracker":self.tracker}),
```
Comment (billonahill) [inline heron/tools/tracker/src/python/main.py]: IIRC we decided that executionstate would be deprecated in favor of /metadata and /runtimestate. Is that correct? If so would you clarify that in the executionstate class, so it's clear to other readers why we have duplicate info returned by multiple endpoints.

Comment (objmagic) [inline heron/tools/tracker/src/python/main.py]: heron-ui is using it. If necessary I will refactor heron-ui in another PR. In the meantime, let us keep this API here.

Comment (objmagic) [inline heron/tools/tracker/src/python/main.py]: imo deprecating execution state could be done in another PR

Comment (billonahill) [inline heron/tools/tracker/src/python/main.py]: I wasn't suggesting removing/refactoring it in this PR, just putting a comment describing that it's slated to be removed. That way if it takes a while to remove people readying the code will understand the duplication.

Comment (objmagic) [inline heron/tools/tracker/src/python/main.py]: Updated: https://github.com/twitter/heron/pull/1734/commits/ee1a73010632b39572e6c53c1299c41545e80637

Also see: https://github.com/twitter/heron/issues/1779

**Comment by billonahill** [comment]
👍

**Pull request merged**